### ISSUE TITLE
Row-wise and column-wise scaling of a sparse matrix runs out of memory

### ISSUE THREAD

**Issue by pitsianis** [JuliaSparse/SparseArrays.jl#543]
As the subject says, when I try to scale the values of a large sparse matrix using `.*`, a dense matrix is built, and `OutOfMemoryError()` is raised!

For instance,
```julia
n = 10_000_000
v = rand(n)
A = sprand(n,n,2/n)
v .* A  # ERROR: OutOfMemoryError()
A .* v' # ERROR: OutOfMemoryError()
```

Using `Diagonal` or `spdiagm` will do the computation, 
```julia
julia> @btime $A * Diagonal($v);
  97.998 ms (6 allocations: 381.50 MiB)

julia> @btime Diagonal($v) * $A;
  134.449 ms (6 allocations: 381.50 MiB)

julia> @btime spdiagm($v) * $A;
  1.794 s (31 allocations: 1.31 GiB)

julia> @btime $A * spdiagm($v);
  522.747 ms (31 allocations: 1.31 GiB)
```
where `spdiagm(v)` forms a sparse diagonal matrix at a great expense in time and space.

See discussion [here](https://discourse.julialang.org/t/scaling-a-sparse-matrix-row-wise-and-column-wise-too-slow/115956)

**Comment by pitsianis** [comment]
This is related to a previous [issue](https://github.com/JuliaSparse/SparseArrays.jl/issues/47)

**Comment by ViralBShah** [comment]
Related: #551

**Comment by pitsianis** [comment]
Isn't it a rational expectation that the sparsity pattern of `A` should not change? If so, `findnz(A)` and only work with them irrespective of the presence of `NaN`s or `Infinity`s?

**Comment by ViralBShah** [comment]
Yes. I will get around to this one with other broadcasting related changes.
